pyrival/algebra/discrete_log.py

```diff
@@ -1,22 +1,50 @@
+from math import gcd, sqrt
+
+
+def extended_gcd(a, b):
+    if not b:
+        return a, 1, 0
+
+    g, y, x = extended_gcd(b, a % b)
+    return g, x, y - ((a // b) * x)
+
+
 def discrete_log(a, b, m):
-    """returns some x s.t. pow(a, x, m) == b or None if no such x exists, works when a amd m are coprime"""
-    a %= m
-    b %= m
-    if b == 1:
+    """returns smallest non-negative x s.t. pow(a, x, m) == b or None if no such x exists"""
+    if b >= m:
+        return None
+
+    if b == 1 or m == 1:
         return 0
 
-    n, vals = int(m**0.5) + 1, {}
-    mult, cur = pow(a, n, m), 1
+    i, t = 1, a
+    while (1 << i) <= m:
+        if t % m == b:
+            return i
+        i, t = i + 1, t * a % m
+
+    raw_b, raw_m = b, m
+    c, d, g = 0, 1, gcd(a, m)
+    while g > 1:
+        b, m = b // g, m // g
+        c, d = c + 1, d * (a // g) % m
+        g = gcd(a, m)
+
+    pos = {}
+    p, n = 1, int(sqrt(m - 0.666)) + 1
```

Comment:
What is this `0.666`? Some reference for this would be nice.

Reply:
Sorry my bad. It's just to avoid precision error. Updating to 0.5 as that will serve the purpose and will make it less alien.

Reply:
Please check now. Should be clearer.